### DISCUSSION TITLE
Two Way Sync

### DISCUSSION THREAD

**Discussion by boatbomber** [rojo-rbx/rojo#728] [Design & Brainstorm]
### What is it?
Two way sync is a mode for Rojo in which the sync goes in both directions: files ⇔ studio. Changes to one propagate to the other.

However, that's just an overall idea. Things get complicated pretty quickly when you start trying to define this feature more concretely. It's (relatively) straightforward for things like script.Source which have a very strict representation on your filesystem as .lua files, but things get fuzzy when syncing other properties or even entire Instances and Models. Should they bake into the project json? As rbxms? Perhaps model jsons? There's a lot of possible paths to take here, and lots of implementation details that need to be carefully thought out and documented before we can tackle creating this killer feature.

### Let's figure this out, together.

This discussion post is here to create a public and searchable space for us to scope out and design a two way sync feature that fills our needs without ambiguity. Once we create our game plan here, we'll be able to open issues for all the pieces that need to be built to make it happen and contributors can get started!

**Comment by u-train** [comment]
To help kick start discussions, here are some of the thoughts I have on what should be done. These are still being worked on, it just sucks that we can't save comments!

**When a script is added**
- If the parent is a directory, add the script to the corresponding file.
- If the parent is an object (model or some blob): TODO
  - We could take the lazy way and just put it as part of the blob, but that wouldn't be very Rojo of us, wouldn't it?
  - I thought there was a way where we could specify in the project _where_ to place something, but that feature isn't there.
  - If it was, we could move the object into a directory with the same name, put the script in there, and make a new project file that describes the structure correctly. We can adjust project files as needed.

**When a script is removed**
- If it's the last file and it's within a directory, delete the script and insert a meta file.
- If it's the last script and the only other significant feature is an object of some kind, move the object out of the directory and delete the directory. Make sure to adjust other project files as needed.
- Otherwise, just simply delete the script and adjust other project files as needed.

**When a new instance is inserted**
- If there are scripts placed deeply, follow the instructions mentioned prior when a script is added and the parent is some object.
- If it's a singular instance, then place a corresponding model file.
- If it's an instance with the majority of ancestors being scripts within it, make a directory, place an `init.meta.json` to describe the instance, and place the scripts following the hierarchy. Place any spare object files as needed.
- If it's an instance with primarily non-scripts, follow the instructions mentioned prior when a script is added and the parent is some object.

**Deciding between a model or blob file**
- Matter of size and how much depth there is.
- It should stick to model files if the hierarchy is shallow and less than 50 instances.
- Otherwise, it should become a rbxmx/rbxmxl file.

**Comment by ryanclkeough** [comment]
Not sure if this project is still being actively worked on, but I just wanted it to be known that I would LOVE this feature.
Currently using Rojo 7.70 on Windows and want Studio to be my main editor (and have Rojo sync back the changes so to track them in GitHub for easier merge requests).
My desired workflow is create/edit things in Studio -> Rojo automatically creates/updates those items -> Git tracks the additions. The existing two-way sync is useful, but creating a new script on the studio side does not sync back as is the point of this thread which at least in my use case nearly defeats the purpose.

I'm happy to test Windows builds or provide reproduction cases.

I don't currently know Rust, but I would be willing to give it a shot if it would help out the project
